### DISCUSSION TITLE
Support for TransactWriteItems / TransactGetItems

### DISCUSSION THREAD

**Discussion by rjmackay** [tywalch/electrodb#163] [Ideas]
Any plans to support transactions?

**Comment by tywalch** [comment]
I have a few things in front of it, but it is definitely on my radar. I also want to make sure the API for it is solid -- right now the ideas I'm playing with aren't clicking with me well. What sort of API would you expect for your use case?

**Reply by estyrke** [reply]
One use case I'm having is to make item creation contingent on some other item existing. Without having any idea what is reasonable to implement, I wouldn't mind something like: 
```typescript
myService.transactWrite((t) => {
  t.entities.myEntity.assertExists.myIndex({ ... });
  t.entities.myOtherEntity.create({...});
});
```

**Reply by rjmackay** [reply]
@tywalch would you accept a PR to add it? Any thoughts on what the function signature should be for it?
i.e. would it be a separate function like `transactWrite` or an execution option on existing methods like `transact: true` ?

**Comment by Easy-Cloud-in** [comment]
My use case is i want to update to entities at the same time. Some thing like this:
```
let response = await TransactionWrite{
    RootUserEntity.update({ })
      .set({  }),
   RootSubscription.update({...})
     .set({})
  .go({),
}
```

**Comment by Easy-Cloud-in** [comment]
Please let us know when this feature landing in ElectroDB. Transaction write  play very important role in development. 
In DynamoDB, Aggregation, Referential Integrity and rest of things can be done using Transaction Writes.

**Comment by maxkondra** [comment]
+1,  having a Transaction Write for multiples items with conditional checks would be great. One good use case of this is:

When adding a new user to a game, we need to:
- Confirm that there are not already 50 players in the game (each game can have a maximum of 50 players).
- Confirm that the user is not already in the game.
- Create a new UserGameMapping entity to add the user to the game.
- Increment the people attribute on the Game entity to track how many players are in the game.

Pulled from https://aws.amazon.com/getting-started/hands-on/data-modeling-gaming-app-with-dynamodb/5/ under "Add users to a game"

```
def join_game_for_user(game_id, username):
    try:
        resp = dynamodb.transact_write_items(
            TransactItems=[
                {
                    "Put": {
                        "TableName": "battle-royale",
                        "Item": {
                            "PK": {"S": "GAME#{}".format(game_id) },
                            "SK": {"S": "USER#{}".format(username) },
                            "game_id": {"S": game_id },
                            "username": {"S": username }
                        },
                        "ConditionExpression": "attribute_not_exists(SK)",
                        "ReturnValuesOnConditionCheckFailure": "ALL_OLD"
                    },
                },
                {
                    "Update": {
                        "TableName": "battle-royale",
                        "Key": {
                            "PK": { "S": "GAME#{}".format(game_id) },
                            "SK": { "S": "#METADATA#{}".format(game_id) },
                        },
                        "UpdateExpression": "SET people = people + :p",
                        "ConditionExpression": "people <= :limit",
                        "ExpressionAttributeValues": {
                            ":p": { "N": "1" },
                            ":limit": { "N": "50" }
                        },
                        "ReturnValuesOnConditionCheckFailure": "ALL_OLD"
                    }
                }
            ]
        )
        print("Added {} to game {}".format(username, game_id))
        return True
    except Exception as e:
        print("Could not add user to game")
```

**Reply by pjvds** [reply]
Transaction support in DynamoDB landed in 2018. This was an important step forward. Many patterns used in DynamoDB use transactions. The book by Alex de Brie describes many well known use cases that can easily be created using DynamoDB, but most of them are based on transactions.

I would love to see this support being added to this great lib. If any help is needed, I'm sure we can donate high quality engineering power.

**Comment by tywalch** [comment]
@pjvds @maxkondra @Easy-Cloud-in @estyrke @rjmackay 

Version `2.5.0` was just shipped with transaction support, here are the docs: https://electrodb.dev/en/mutations/transact-write/

**Reply by estyrke** [reply]
Awesome! I just tried it out, feels great! Thanks for all the effort!

**Comment by pjvds** [comment]
Thanks for the great work! This is really helpful for us. <3

**Comment by maxkondra** [comment]
Super awesome, thank you Tyler! Looking forward to trying it out.
